src/WebJobs.Script.WebHost/Routing/ScriptRouteHandler.cs

```diff
@@ -37,8 +42,25 @@ public async Task InvokeAsync(HttpContext context, string functionName)
                 ProxyFunctionExecutor proxyFunctionExecutor = new ProxyFunctionExecutor(_scriptHost);
                 context.Items.TryAdd(ScriptConstants.AzureProxyFunctionExecutorKey, proxyFunctionExecutor);
             }
+            else if (_isWarmup)
+            {
+                // warmup function will get executed just once for the process.
+                if (Interlocked.CompareExchange(ref _warmupExecuted, 1, 0) != 0)
```

Comment:
You're maintaining the logic that existed in the controller method I see. @ankitkumarr, why do we need to enforce a single execution? I'd say if someone calls again for some reason, we just invoke again.

Reply:
@mathewc again nobody can call warmup functions from outside, they are not http triggers. the idea is user adds all initialization of heavy objects and what not in this warmup function and it only gets called ONCE for each process. Also FEs and Data roles will call this /admin/warmup end point to make sure host is up and running. FE by default will call it every few minutes as part of AlwaysOn and we need to make sure the call is very light and returns quickly except the very first time that we do initializations. This behavior of warmup functions is documented as well.

Reply:
@ankit - when customers develop/test their warmup function, they presumably use portal/cli to invoke/test the function. If the existing behavior is to only run it once, they can only run/debug once, then they have to restart their app to try again? Seems artificial to me.

Reply:
Not sure if anyone has ever complained about invoke/testing of this but if it's ok to break the existing behavior, maybe we could allow this only once if environment is app service and otherwise remove this requirement for cli scenarios and local testing and debugging. would still fail for portal but works for cli.

Reply:
> why do we need to enforce a single execution? I'd say if someone calls again for some reason, we just invoke again.

What Hamid said and also when we built this, there was an issue where sometimes platform called `/admin/warmup` multiple times during startup. To avoid users incurring costs for execution and avoiding duplicate calls, we added this.

> when customers develop/test their warmup function, they presumably use portal/cli to invoke/test the function. If the existing behavior is to only run it once, they can only run/debug once, then they have to restart their app to try again? Seems artificial to me.

Yeah that makes sense. Right now, you can't even execute once outside of Azure as we have a check to only run when in Elastic Premium Azure environment. I think that we can expose a way to run locally and execute it. Probably can take that separately (in a different PR) as that would be a separate discussion.